### ISSUE TITLE
Storage node is unable to use multiple cores.

### ISSUE THREAD

**Issue by Pentium100MHz** [storj/storj#2373]
During the latest tests I noticed that, while all cores are utilized, the combined CPU usage does not go above 110%, indicating that there is something that is single threaded (and all threads take turns in accessing it).
In addition, this slows down with the amount of data stored on the node.

Average detailed CPU usage (maximum 100%)
![image](https://user-images.githubusercontent.com/8972050/60324609-e8ff6180-998d-11e9-8ba8-d537160df1a3.png)
A very clear flat top. The jump from 12% to 25% happened because I reduced the core count in the VM from 8 to 4. 

Traffic during that time (decreasing as the amount of data increases):
![image](https://user-images.githubusercontent.com/8972050/60324712-2b28a300-998e-11e9-9061-61e455270ed4.png)

CPU utilization per core, stacked (maximum would be 400%), shorted timespan because I created this graph later:
![image](https://user-images.githubusercontent.com/8972050/60324797-67f49a00-998e-11e9-888d-c6c2f185f9a1.png)
Again, a very clear flat top at 110%.

Traffic during the time span of the indivitual core graph:
![image](https://user-images.githubusercontent.com/8972050/60324974-d3d70280-998e-11e9-9308-82b3d3261185.png)

**Comment by kaloyan-raev** [comment]
@Pentium100MHz Thank for these graphs. They are very valuable to us!

Could you give some details about the hardware?

**Comment by Pentium100MHz** [comment]
The node is running inside a VM that is running on a server with 2x Xeon X5687 CPUs (4cores/8threads each) and 100GB RAM. Its single core performance is not super fast, but the fastest CPU, AMD Ryzen 5 3600 scores only twice as much, according to cpubenchmark.net

The VM has 16GB RAM and 4 cores (8 cores before) assigned to it. There are a couple of other VMs running on that server, but they do not create any noticeable CPU load.

Virtualization platform is KVM, the storage is ext4 (inside the VM) in a zvol (on the host). The array has SSD cache for reads (L2ARC) and writes (ZIL). The array is RAIDz2 using 6 drives (was RAID1 from 2 drives before).
I found a few other graphs you may find interesting.

The graphs are taken from the VM.
This is the CPU usage graph from the host :
![image](https://user-images.githubusercontent.com/8972050/60327525-c0c73100-9994-11e9-88ef-2cd7d33434c0.png)

The jump above 10% is because during that time I was moving data from the old array to the new one. As you can see in the traffic graph, it did not affect the traffic at all. Also note that IO wait is less than 2.5%, well, except the time when I moved the data.

This is the Unix load of the host:
![image](https://user-images.githubusercontent.com/8972050/60327712-4a76fe80-9995-11e9-8bb1-2d9596e14c9a.png)
Again, a horizontal line, except the time when I moved the data.

Disk IO load as seen from the VM:
![image](https://user-images.githubusercontent.com/8972050/60327794-7eeaba80-9995-11e9-87a8-c15e86b93da3.png)
One spike is because I tried forcing sync writes to see how it affects performance, the spike in June 24th is when the node started up from the new array for the first time (and, as such, did not have anything cached).

Disk IO bytes/sec as seen from the VM:
![image](https://user-images.githubusercontent.com/8972050/60327987-ee60aa00-9995-11e9-9e81-0d4701bcc91a.png)

The spike for writes was during the time I enabled sync writes, the spike for reads was the node startup. Other than that, pretty much all reads came from the cache inside the VMs 16GB of RAM.

**Comment by kaloyan-raev** [comment]
@Pentium100MHz Thank you for the additional info!

Could you check if this CPU usage is indeed consumed by the `storagenode` process, and not by another process on the VM?

I talked to some folks at Storj who regularly monitor their storage nodes. They claim they haven't seen such behavior.

**Comment by Pentium100MHz** [comment]
Yes, storagenode process takes the CPU.
```
  1  [||||||||||||||||||||||||||||||||||||||||||||||                                 53.0%]   3  [|||                                                                             2.6%]
  2  [||||||                                                                          5.4%]   4  [||||||||||||||||||||||||||||||||||||||||||||                                   51.0%]
  Mem[|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||1.01G/15.7G]   Tasks: 41, 121 thr; 2 running
  Swp[|                                                                          56K/1022M]   Load average: 0.63 0.72 0.70 
                                                                                              Uptime: 4 days, 04:55:36

  PID USER      PRI  NI  VIRT   RES   SHR S CPU% MEM%   TIME+  Command
20683 root       20   0  866M  752M 17264 S 110.  4.7 23h37:39 ./storagenode run --config-dir config --identity-dir identity --metrics.app-suffix=-alpha --metrics.interval=30m --kademlia.ex
20807 root       20   0  866M  752M 17264 S 67.4  4.7  1h33:30 ./storagenode run --config-dir config --identity-dir identity --metrics.app-suffix=-alpha --metrics.interval=30m --kademlia.ex
20729 root       20   0  866M  752M 17264 R 34.7  4.7  1h51:36 ./storagenode run --config-dir config --identity-dir identity --metrics.app-suffix=-alpha --metrics.interval=30m --kademlia.ex
20731 root       20   0  866M  752M 17264 S  3.3  4.7 45:10.66 ./storagenode run --config-dir config --identity-dir identity --metrics.app-suffix=-alpha --metrics.interval=30m --kademlia.ex
20733 root       20   0  866M  752M 17264 S  2.0  4.7  1h58:54 ./storagenode run --config-dir config --identity-dir identity --metrics.app-suffix=-alpha --metrics.interval=30m --kademlia.ex
 8177 root       20   0  866M  752M 17264 S  1.3  4.7  1h35:21 ./storagenode run --config-dir config --identity-dir identity --metrics.app-suffix=-alpha --metrics.interval=30m --kademlia.ex
22786 root       20   0  866M  752M 17264 S  0.7  4.7  1h48:16 ./storagenode run --config-dir config --identity-dir identity --metrics.app-suffix=-alpha --metrics.interval=30m --kademlia.ex
20718 root       20   0  866M  752M 17264 S  0.7  4.7 19:56.43 ./storagenode run --config-dir config --identity-dir identity --metrics.app-suffix=-alpha --metrics.interval=30m --kademlia.ex
```
Sometimes one of the threads goes to 100%, while other threads stay at 5% or so. It is not the same thread every time.

For example:
```
  1  [||                                                                              1.3%]   3  [|                                                                               0.7%]
  2  [||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||100.0%]   4  [||||                                                                            3.4%]
  Mem[|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||1.01G/15.7G]   Tasks: 41, 121 thr; 2 running
  Swp[|                                                                          56K/1022M]   Load average: 0.96 0.82 0.74 
                                                                                              Uptime: 4 days, 04:58:32

  PID USER      PRI  NI  VIRT   RES   SHR S CPU% MEM%   TIME+  Command
20683 root       20   0  867M  756M 21404 S 106.  4.7 23h39:34 ./storagenode run --config-dir config --identity-dir identity --metrics.app-suffix=-alpha --metrics.interval=30m --kademlia.ex
21294 root       20   0  867M  756M 21404 R 100.  4.7  1h46:16 ./storagenode run --config-dir config --identity-dir identity --metrics.app-suffix=-alpha --metrics.interval=30m --kademlia.ex
 8177 root       20   0  867M  756M 21404 S  1.3  4.7  1h35:24 ./storagenode run --config-dir config --identity-dir identity --metrics.app-suffix=-alpha --metrics.interval=30m --kademlia.ex
20733 root       20   0  867M  756M 21404 S  1.3  4.7  1h59:19 ./storagenode run --config-dir config --identity-dir identity --metrics.app-suffix=-alpha --metrics.interval=30m --kademlia.ex
20731 root       20   0  867M  756M 21404 S  1.3  4.7 45:21.00 ./storagenode run --config-dir config --identity-dir identity --metrics.app-suffix=-alpha --metrics.interval=30m --kademlia.ex
20718 root       20   0  867M  756M 21404 S  1.3  4.7 19:58.48 ./storagenode run --config-dir config --identity-dir identity --metrics.app-suffix=-alpha --metrics.interval=30m --kademlia.ex
```

The "maximum" thread, other than 20683 which is the main process and shows the sum, changes every time htop refreshes, the numbers also jump around. But sometimes it pegs one core to the top for a brief period. This happened more frequently during the long test obviously.

The OS is Debian 9 (both VM and host), docker version:
```
Client:
 Version:           18.09.6
 API version:       1.39
 Go version:        go1.10.8
 Git commit:        481bc77
 Built:             Sat May  4 02:36:00 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.6
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.8
  Git commit:       481bc77
  Built:            Sat May  4 01:59:36 2019
  OS/Arch:          linux/amd64
  Experimental:     false
```

**Comment by gsxryan** [comment]
I can confirm the same behavior, during the same stress test on a 4 core 3.2Ghz i3.  I'll leave some more helpful info after work.

![image](https://user-images.githubusercontent.com/2055136/60341224-6b2e5c80-9973-11e9-930a-631e48962105.png)

**Comment by gsxryan** [comment]
I don't have time to test this today, but perhaps we should look into docker run settings for CPU pinning, or CPU limit.

For X CPUs:
Where Limit = X00%
Or CPU Pinning for 1,2,3 (for 4 CPUs), omit 0 for system.

**Comment by JL421** [comment]
So I have tested pulling the storagenode executable out of the Docker container and running it. It does show the same behavior outside as inside the container.

**Comment by gsxryan** [comment]
It seems JL's test eliminates my suspicion.  If it's the same outside of docker, it's surely a software issue, i'm not sure how you replicated it so quickly / reliably though.  I rarely see the condition.  

Also, to help designate a priority on this issue, here's my view from a macro perspective.  Regarding StorJ's announcement about IP filtering.

> IP filtering: There is no advantage to have multiple nodes on the same IP address or subnet. If you want to run more than one node, you can, but you’ll still get the same amount of data as both nodes would get combined. Our network design requires strong nodes that are geographically distributed—we do this in case a specific area is impacted, we’re able to reconstitute our users' files.

Running in the state in this issue actually bottlenecks at the CPU.  SNOs running multiple nodes will actually benefit in this condition, so while each node will get less traffic, it will bottleneck less often than SNO's with only a single node.

Example,
Running 4 cores, 3 nodes in the condition in this github issue will eventually result in the same DoS like traffic, but will split to each node (110% each) for a total of 330% CPU usage (out of 400% possible).  This greatly increases the chance you will receive data, and not get a constant stream of context cancelled messages due to an unresponsive (maxed out CPU node).  This SNO will get more traffic than one who only can max their CPU at 110% vs the SNO that can max CPU at 330% with 3 nodes.

**Comment by JL421** [comment]
I had tested it last week when Pentium was also running his tests.

**Comment by gsxryan** [comment]
Were these users running the RPI3s?  Or, more powerful hardware?   Because of so many reports of RPIs crashing, I suspect that the RPIs are crashing in the very early part of the condition in this issue.  Then they reboot, it fixes itself.  The low power hardware increases the chances of it falling into this state (speculative of course)

> I talked to some folks at Storj who regularly monitor their storage nodes. They claim they haven't seen such behavior.

**Comment by stefanbenten** [comment]
@Pentium100MHz @gsxryan @JL421 

**TLDR:** The Software is totally able to use multiple cores, the limit here is _currently_ the SQLite Database, that blocks the other requests.

For comparsion, more in-depth explanation:
My Storagenode is running on a VM with 4 Cores, 8GB RAM, on top of a VMware Cluster.

The CPUs running on these hosts are Intel Gold 6154 (18C, 3.0GHz).
The Storage of that VM is based on a ZFS SAN, that uses SATA SSDs (Intel S3610) for raw storage, together a pair of Intel P4800X as ZIL.

One of the limiting factors right now is the SQLite database. Especially on slower machines, it is bottleneck and causes the the load to spike to 1. Currently that is due to entire table locks, that the team is working on getting removed asap.
That said, my entire (base) VM is being kept in ARC, and the storage location is living on a second "drive". While the uploads i did over the past month, my node was happily distributing the load across all 4 cores, where some of them spiked as well to 70-80% shortly.

Also, we are not stating everything is perfect, thats why we are still in Alpha! We value the information very much and try to fix bottlenecks with highest priority, to leverage our network to the next step/level 👍 
Once the Fix for the Database is merged and released, i will let you guys now and would appreciate the feedback then!

**Comment by gsxryan** [comment]
Perfect Stefan!  Happy to see it is known and being worked on.  Will keep following to see when to test next.

> the limit here is currently the SQLite Database, that blocks the other requests.

> Currently that is due to entire table locks, that the team is working on getting removed asap.

**Comment by Pentium100MHz** [comment]
Good to know that this is being worked on and will be fixed in the future.

**Comment by stefanbenten** [comment]
@gsxryan Not really. I also tested that in a dev environment.

**Comment by Pentium100MHz** [comment]
A new observation: seems that the database slows down primarily due to the accumulated ingress/egress traffic and not the amount of data stored as the performance of my node jumped up after the counterswere reset:
![image](https://user-images.githubusercontent.com/8972050/60409762-97ddb080-9bcd-11e9-99d9-41e0e020d665.png)
![image](https://user-images.githubusercontent.com/8972050/60409797-b643ac00-9bcd-11e9-9324-576d105fe8cb.png)

Still, making the database multi-threaded will fix this.

**Comment by gsxryan** [comment]
![Point Blur_Jun302019_232151](https://user-images.githubusercontent.com/2055136/60410528-65f92980-9b8e-11e9-853a-b4e7f2eba54c.jpg)

I noticed an even bigger throughput difference. Seemed like very high CPU usage at the end of the month, then dropoff maintaining same throughput, and shortly after a big increase in xfer.

**Comment by venadCLI** [comment]
Excuse for, but which software you use for monitor?

**Comment by gsxryan** [comment]
> Excuse for, but which software you use for monitor?

In order of appearance there are a few technologies in use for monitoring in this thread:
Zabbix
Cacti
Telegraf/InfluxDB/Grafana

**Comment by stefanbenten** [comment]
@Pentium100MHz and everyone else.
From my tests, this issue is being resolved. For that reason i am going to close the issue.

If anyone encounters this issue is still persistent, please reopen!
